librz/core/cmd_debug.c

```diff
@@ -1443,376 +1436,408 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
-				if (sectname) {
-					char *sect = rz_str_escape(sectname);
-					res = rz_sys_cmd_strf("env RZ_BIN_PREFIX=\"%s\" rz-bin %s-B 0x%08" PFMT64x " -S \"%s\" | grep \"%s\"", name, mode, baddr, filesc, sect);
-					free(sect);
-				} else {
-					res = rz_sys_cmd_strf("env RZ_BIN_PREFIX=\"%s\" rz-bin %s-B 0x%08" PFMT64x " -S \"%s\"", name, mode, baddr, filesc);
+				char *newfile = NULL;
+				if (!rz_file_exists(file)) {
+					newfile = rz_file_temp("memlib");
+					if (newfile) {
+						file = newfile;
+						rz_core_cmdf(core, "wtf %s 0x%" PFMT64x " @ 0x%" PFMT64x " 2> %s",
+							file, map->size, baddr, RZ_SYS_DEVNULL);
+					}
 				}
-				free(filesc);
-				rz_cons_println(res);
-				free(name);
-				free(res);
-				if (libname || addr != UT64_MAX) { //only single match requested
-					break;
+				get_bin_info(core, file, baddr, pj, mode, symbols_only, &filter);
+				if (newfile) {
+					if (!rz_file_rm(newfile)) {
+						eprintf("Error when removing %s\n", newfile);
+					}
+					free(newfile);
 				}
+			} else {
+				rz_bin_set_baddr(core->bin, map->addr);
+				rz_core_bin_info(core, RZ_CORE_BIN_ACC_SYMBOLS, pj, (input[0] == '*'), true, &filter, NULL);
+				rz_bin_set_baddr(core->bin, baddr);
 			}
 		}
+		if (mode == RZ_MODE_JSON) {
+			rz_cons_println(pj_string(pj));
+			pj_free(pj);
+		}
 		free(ptr);
 	} break;
-	case ' ': // "dm "
+	case '.': // "dmi."
 	{
-		int size;
-		char *p = strchr(input + 2, ' ');
-		if (p) {
-			*p++ = 0;
-			addr = rz_num_math(core->num, input + 1);
-			size = rz_num_math(core->num, p);
-			rz_debug_map_alloc(core->dbg, addr, size, false);
-		} else {
-			eprintf("Usage: dm addr size\n");
-			return false;
+		map = get_closest_map(core, addr);
+		if (map) {
+			ut64 closest_addr = UT64_MAX;
+			RzList *symbols = rz_bin_get_symbols(core->bin);
+			RzBinSymbol *symbol, *closest_symbol = NULL;
+
+			rz_list_foreach (symbols, iter, symbol) {
+				if (symbol->vaddr > addr) {
+					if (symbol->vaddr - addr < closest_addr) {
+						closest_addr = symbol->vaddr - addr;
+						closest_symbol = symbol;
+					}
+				} else {
+					if (addr - symbol->vaddr < closest_addr) {
+						closest_addr = addr - symbol->vaddr;
+						closest_symbol = symbol;
+					}
+				}
+			}
+			if (closest_symbol) {
+				RzCoreBinFilter filter;
+				filter.offset = 0LL;
+				filter.name = (char *)closest_symbol->name;
+
+				rz_bin_set_baddr(core->bin, map->addr);
+				rz_core_bin_info(core, RZ_CORE_BIN_ACC_SYMBOLS, NULL, false, true, &filter, NULL);
+			}
 		}
 	} break;
-	case '-': // "dm-"
-		if (input[1] != ' ') {
-			eprintf("|ERROR| Usage: dm- [addr]\n");
-			break;
-		}
-		addr = rz_num_math(core->num, input + 2);
+	default:
+		rz_core_cmd_help(core, help_msg_dmi);
+		break;
+	}
+	return RZ_CMD_STATUS_OK;
+}
+
+// dmp
+RZ_IPI RzCmdStatus rz_cmd_debug_dmp_handler(RzCore *core, int argc, const char **argv) {
+	RzListIter *iter;
+	RzDebugMap *map;
+	ut64 addr = 0, size = 0;
+	int perms;
+	if (argc == 4) { // dmp <addr> <size> <perms>
```

Comment:
these checks are not in sync with the optional/non-optional arguments defined in the .yaml. In particular, I see `dmp [<addr> [<size> <perms>]]`, but if i do `dmp` it says `Wrong number of arguments passed to `dmp`, see its help with `dmp?``.

Reply:
`dmp` commands can be run in two ways:  
- dmp <addr> <size> <perms>
- dmp <perms> 
How should the yaml file look like for input like this?

Reply:
let's change the way this command works IMHO. @XVilka opinions?

For example `dmp <perms> [<size>]` (again, I would not have `<addr>` in the arguments because that can be changed with `@ <addr>`). WDYT?

Reply:
Yeah, this seems like a nice way to deal with the case.